### ISSUE TITLE
Ubuntu 14.04, Platform Haskell 2014.02 issue on fresh new installation

### ISSUE THREAD

**Issue by varosi** [JPMoresmau/eclipsefp#254]
Fresh configuration:
* Ubuntu 14.04
* Platform Haskell 2014.02
* EclipseFP (latest)
* Fresh new workspace

1. Open Eclipse
2. It update from internet latest cabal info
3. Starts installation of different tools automatically
4. After some time of waiting I see this in my console window:
> Resolving dependencies...
> cabal: Could not resolve dependencies:
> trying: HTTP-4000.2.19/installed-230... (user goal)
> trying: aeson-0.8.0.2/installed-dd9... (user goal)
> next goal: parsec (user goal)
> rejecting: parsec-3.1.7/installed-4e6... (conflict: HTTP =>
> parsec==3.1.5/installed-d6d...)
> rejecting: parsec-3.1.5/installed-d6d... (conflict: aeson =>
> text==1.2.0.3/installed-de9..., parsec => text==1.1.0.0/installed-155...)
> rejecting: parsec-3.1.7, 3.1.6, 3.1.5, 3.1.4, 3.1.3, 3.1.2, 3.1.1, 3.1.0,
> 3.0.1, 3.0.0, 2.1.0.1, 2.1.0.0, 2.0 (conflict: HTTP =>
> parsec==3.1.5/installed-d6d...)
> Backjump limit reached (change with --max-backjumps).
> 
> Note: when using a sandbox, all packages are required to have consistent
> dependencies. Try reinstalling/unregistering the offending packages or
> recreating the sandbox.

On fresh new cabal sandbox I have succeeded with SourceGraph installation. So I think previous tools that are installed have dependencies which conflict with SourceGraph dependencies.

I think one solution could be to always install those tools with "--force-reinstalls". They are installed in a sandbox anyway.

Or each tool to be installed in its own sandbox.

**Comment by varosi** [comment]
Another problem is that when it failed with SourceGraph it didn't install Hoogle at all. As I understood those tools they are not so related, so braking one of them should not stop other from building.

**Comment by JPMoresmau** [comment]
Yes in a sandbox you have to have one consistent set of packages. If we install all tools at once, all may fail and you end up with no executables. If we install one by one, you always get the first ones but some later tool may fail even though they could have installed together. If you install each tool in a different sandbox, you install the same base package n times and waste a lot of disk space and CPU cycles.

**Comment by varosi** [comment]
If they are built in separate sandboxes sequentially it will not cost more disk space than a single build. 
May be some hybrid version could do the job in a stable manner like:
1. Check what tools shall be built. Don't build already built healthy tools (tool --version works for example).
2. Build sequentially those tools in a single sandbox. 
3. If some of them fail, delete this sandbox and continue with new one from the place where first broken.
4. Go to 2 until passed all the tools.

**Comment by JPMoresmau** [comment]
"If they are built in separate sandboxes sequentially it will not cost more disk space than a single build." I don't understand. If each tool is in a separate sandbox, then each library they depend on will be installed once in each sandbox, so you'll have more disk usage.
Having several sandboxes is going to be a nightmare, but I suppose to be really solid we would have to:
1) build all tools together at once to try to have all dependencies done together
2) if this fails, try to build all tools one by one in the same sandbox, starting in order of importance (buildwrapper first, etc)
3) tools that fail can be built in their own sandbox.

But already in the dev version I've removed properly SourceGraph.

I'm not sure --version is supported for each tool we use, which add to the confusion...

**Comment by varosi** [comment]
{some_tool} --version 
will run the tool to see its version. If it's running well and displaying a version then it works. If not then - it's not working.

About separate sandboxes I mean to:
1. cabal sandbox init
2. cabal install buildwrapper
3. copy buildwrapper.exe to somewhere
4. cabal sandbox delete
5. goto 1
This will take constant disk space.

SourceGraph could be built the same way. So if it only fails everything else will be built anyway. And on next EclipseFP run it will try to build only SourceGraph and not try to rebuild everything.

**Comment by JPMoresmau** [comment]
is that now a duplicate of #220 ?

**Comment by varosi** [comment]
Yes, I think the problem is the same. But in this issue we have more discussion about the problem.

**Comment by neothemachine** [comment]
I just tried installed the helpers manually:
```
$ cabal install buildwrapper scion-browser hoogle stylish-haskell hlint
Resolving dependencies...
cabal: Could not resolve dependencies:
trying: buildwrapper-0.9.1 (user goal)
trying: conduit-extra-1.1.6.2 (dependency of buildwrapper-0.9.1)
trying: streaming-commons-0.1.9.1 (dependency of conduit-extra-1.1.6.2)
trying: random-1.1 (dependency of streaming-commons-0.1.9.1)
trying: aeson-0.8.0.2 (dependency of buildwrapper-0.9.1)
trying: template-haskell-2.9.0.0/installed-6d2... (dependency of
aeson-0.8.0.2)
trying: scion-browser-0.5.0 (user goal)
next goal: parallel-io (dependency of scion-browser-0.5.0)
rejecting: parallel-io-0.3.3, 0.3.2.2, 0.3.2.1 (conflict: random==1.1,
parallel-io => random>=1.0 && <1.1)
rejecting: parallel-io-0.3.2, 0.3.1, 0.3.0.2, 0.3.0.1 (conflict:
template-haskell => containers==0.5.5.1/installed-d4b..., parallel-io =>
containers>=0.2 && <0.5)
rejecting: parallel-io-0.3.0 (conflict: template-haskell =>
containers==0.5.5.1/installed-d4b..., parallel-io => containers>=0.3 && <0.4)
rejecting: parallel-io-0.2.2, 0.2.1.1, 0.2.1, 0.2 (conflict: scion-browser =>
parallel-io>=0.3)
Backjump limit reached (change with --max-backjumps).
```

I guess multiple sandboxes are really the way to go.

**Comment by neothemachine** [comment]
Ha, funny, with `--reorder-goals` it works.
